src/requestManager.cpp

```diff
@@ -1105,41 +1186,80 @@ bool CRequestManager::MarkBlockAsReceived(const uint256 &hash, CNode *pnode)
                 }
             }
         }
-        // BUIP010 Xtreme Thinblocks: end section
-        if (state->vBlocksInFlight.begin() == itInFlight->second.second)
+
+        if (state->vBlocksInFlight.begin() == mapBlocksInFlight[hash][nodeid])
         {
             // First block on the queue was received, update the start download time for the next one
             state->nDownloadingSince = std::max(state->nDownloadingSince, GetTimeMicros());
         }
-        state->vBlocksInFlight.erase(itInFlight->second.second);
+        state->vBlocksInFlight.erase(mapBlocksInFlight[hash][nodeid]);
         state->nBlocksInFlight--;
-        mapBlocksInFlight.erase(itInFlight);
+        MapBlocksInFlightErase(hash, nodeid);
+
         return true;
     }
     return false;
 }
 
+void CRequestManager::MapBlocksInFlightErase(const uint256 &hash, NodeId nodeid)
+{
+    // If there are more than one block in flight for the same block hash then we only remove
+    // the entry for this particular node, otherwise entirely remove the hash from mapBlocksInFlight.
+    LOCK(cs_objDownloader);
+    if (mapBlocksInFlight.count(hash) && mapBlocksInFlight[hash].size() > 1)
+    {
+        mapBlocksInFlight[hash].erase(nodeid);
+    }
+    else
+    {
+        mapBlocksInFlight.erase(hash);
+    }
```

Comment:
In case we have only one block in flight for a particular hash wouldn't ` mapBlocksInFlight[hash].erase(nodeid)` be equal to `mapBlocksInFlight.erase(hash)`  ?

Reply:
thanks, done...